realtime/ash/ash_event.py

```diff
@@ -0,0 +1,814 @@
+        extent = impact_function.hazard.extent()
+        hazard_extent = [
+            extent.xMinimum(), extent.yMinimum(),
+            extent.xMaximum(), extent.yMaximum()]
+
+        # clip exposure if required (if it is too large)
+        if isinstance(exposure_layer, QgsRasterLayer):
+            cell_size, _ = get_wgs84_resolution(exposure_layer)
+        else:
+            cell_size = None
+        clipped_exposure = clip_layer(
+            layer=exposure_layer,
+            extent=hazard_extent,
+            cell_size=cell_size)
+        exposure_layer = clipped_exposure
+
+        impact_function.exposure = exposure_layer
+        impact_function.requested_extent = hazard_extent
+        impact_function.requested_extent_crs = impact_function.hazard.crs()
+        impact_function.force_memory = True
+
+        try:
+            impact_function.run_analysis()
+            impact_layer = impact_function.impact
+
+            if impact_layer:
+                self.set_impact_style(impact_layer)
+
+                # copy results of impact to report_path directory
+                self.copy_layer(impact_layer, output_basename)
+        except ZeroImpactException as e:
+            # in case zero impact, just return
+            LOGGER.info('No impact detected')
+            LOGGER.info(e.message)
+            return False
+        except Exception as e:
+            LOGGER.info('Calculation error')
+            LOGGER.exception(e)
+            return False
+        LOGGER.info('Calculation completed.')
+        return True
+
+    def calculate_impact(self):
+        # calculate population impact
+        LOGGER.info('Calculating Impact Function')
+        population_impact_success = self.calculate_specified_impact(
+            'AshRasterPopulationFunction',
+            self.hazard_layer,
+            self.population_layer,
+            'population_impact')
+
+        # calculate landcover impact
+        landcover_impact_success = self.calculate_specified_impact(
+            'AshRasterLandCoverFunction',
+            self.hazard_layer,
+            self.landcover_layer,
+            'landcover_impact')
+
+        # calculate cities impact
+        cities_impact_success = self.calculate_specified_impact(
+            'AshRasterHazardPlacesFunction',
+            self.hazard_layer,
+            self.cities_layer,
+            'cities_impact')
+
+        # calculate airport impact
+        airport_impact_success = self.calculate_specified_impact(
+            'AshRasterHazardPlacesFunction',
+            self.hazard_layer,
+            self.airport_layer,
+            'airport_impact')
+        self.impact_exists = True
+
+    def generate_report(self):
+        # Generate pdf report from impact/hazard
+        LOGGER.info('Generating report')
+        if not self.impact_exists:
+            # Cannot generate report when no impact layer present
+            LOGGER.info('Cannot Generate report when no impact present.')
+            return
```

Comment:
Do we have any criteria for this?
Or we can assume that we should always generate the report?